sticker/src/encoder/categorical/mod.rs

```diff
@@ -1,44 +1,122 @@
 use std::hash::Hash;
+use std::marker::PhantomData;
 
 use conllx::graph::Sentence;
 use failure::Error;
 
 use crate::encoder::{EncodingProb, SentenceDecoder, SentenceEncoder};
 use crate::Numberer;
 
+mod mutability {
+    use std::cell::RefCell;
+    use std::hash::Hash;
+
+    use crate::Numberer;
+
+    pub trait Mutability<V>
+    where
+        V: Eq + Hash,
+    {
+        fn new(numberer: Numberer<V>) -> Self;
+
+        fn number(&self, value: V) -> Option<usize>;
+
+        fn value(&self, number: usize) -> Option<V>;
+    }
+
+    pub struct Immutable<V>(Numberer<V>)
+    where
+        V: Eq + Hash;
+
+    impl<V> Mutability<V> for Immutable<V>
+    where
+        V: Clone + Eq + Hash,
+    {
+        fn new(numberer: Numberer<V>) -> Self {
+            Immutable(numberer)
+        }
+
+        fn number(&self, value: V) -> Option<usize> {
+            self.0.number(&value)
+        }
+
+        fn value(&self, number: usize) -> Option<V> {
+            self.0.value(number).cloned()
+        }
+    }
+
+    pub struct Mutable<V>(RefCell<Numberer<V>>)
+    where
+        V: Eq + Hash;
+
+    impl<V> Mutability<V> for Mutable<V>
+    where
+        V: Clone + Eq + Hash,
+    {
+        fn new(numberer: Numberer<V>) -> Self {
+            Mutable(RefCell::new(numberer))
+        }
+
+        fn number(&self, value: V) -> Option<usize> {
+            Some(self.0.borrow_mut().add(value))
+        }
+
+        fn value(&self, number: usize) -> Option<V> {
+            self.0.borrow().value(number).cloned()
+        }
+    }
```

Comment:
Maybe `MutableNumberer` and `ImmutableNumberer` would be more descriptive names than `Mutable` and `Immutable`.  Perhaps also rename `Mutability` to something like `Number` since this seems to be the primary function of the trait.

Then we would have a trait `Number` implemented on `MutableNumberer` and `ImmutableNumberer`

Reply:
Sounds good.